iothub/service/src/FileUpload/FileUploadNotificationProcessorClient.cs

```diff
@@ -0,0 +1,215 @@
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Disposing FileUploadNotificationProcessorClient", nameof(Dispose));
+
+            _fileNotificationReceiver.Dispose();
+
+            if (Logging.IsEnabled)
+                Logging.Exit(this, $"Disposing FileUploadNotificationProcessorClient", nameof(Dispose));
+            GC.SuppressFinalize(this);
+        }
+
+        private async void OnNotificationMessageReceivedAsync(AmqpMessage amqpMessage)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, amqpMessage, nameof(OnNotificationMessageReceivedAsync));
+
+            try
+            {
+                if (amqpMessage != null)
+                {
+                    using (amqpMessage)
+                    {
+                        AmqpClientHelper.ValidateContentType(amqpMessage, CommonConstants.FileNotificationContentType);
+                        FileUploadNotification fileUploadNotification = await AmqpClientHelper.GetObjectFromAmqpMessageAsync<FileUploadNotification>(amqpMessage).ConfigureAwait(false);
+                        fileUploadNotification.DeliveryTag = amqpMessage.DeliveryTag;
+                        AcknowledgementType ack = FileUploadNotificationProcessor.Invoke(fileUploadNotification);
+                        switch (ack)
+                        {
+                            case AcknowledgementType.Abandon:
+                                await _fileNotificationReceiver.AbandonAsync(fileUploadNotification, CancellationToken.None).ConfigureAwait(false);
+                                break;
+
+                            case AcknowledgementType.Complete:
+                                await _fileNotificationReceiver.CompleteAsync(fileUploadNotification, CancellationToken.None).ConfigureAwait(false);
+                                break;
+
+                            default:
+                                break;
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(OnNotificationMessageReceivedAsync)} threw an exception: {ex}", nameof(OnNotificationMessageReceivedAsync));
+                if (ex is IotHubException || ex is IOException)
+                {
+                    if (ex is IotHubException)
+                        ErrorProcessor?.Invoke(new ErrorContext((IotHubException)ex));
+                    else
+                        ErrorProcessor?.Invoke(new ErrorContext((IOException)ex));
+                }
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, amqpMessage, nameof(OnNotificationMessageReceivedAsync));
+            }
+        }
+
+        private void ConnectionClosed(object sender, EventArgs e)
+        {
+            var ex = new IotHubException(e.ToString());
```

Comment:
This isn't all that helpful. Ideally, we would parse out any AMQP level error code so that users can interpret this exception. Sections 2.8.16 - 2.8.18 of the [AMQP spec](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-complete-v1.0-os.pdf) enumerate the possible error codes we should expect from the AMQP client here.

Some of these errors are network level issues (and should be given to users as IOExceptions) such as "amqp:connection:framing-error" whereas some errors are Hub level issues such as "amqp:link:message-size-exceeded" when a user sends a message that is too big.

Bryce is working on the broad exception strategies for our v2 SDK, so I'm not asking you to figure that out right now, but I would expect some kind of parsing logic to get the AMQP error code. The device client side should be doing this already if you want to take reference from it.

Reply:
Will the exception be passed as an instance of EventArgs or I would need to convert it into an instance of amqp exception?

Reply:
@timtay-microsoft added the requested changes.